Determine non-process name lookups at begin job

Dr15Jones opened the pull request:
#### PR description:

When a specific process name is not given during the `consumes` call, the system now finds the best matching process to use at begin job time. If at transition processing time the data product matched is not found, no further matches are attempted.

Internally the ProductResolverIndexHelper now holds an entry to lookup which ProductResolver to call if one requests to  _skip current process_ . This is now also determined at begin job time rather than at run time.

#### PR validation:

Code compiles and all framework unit tests (after modification) pass.

resolves https://github.com/cms-sw/framework-team/issues/1522

Dr15Jones committed:
Better process order handling

Dr15Jones committed:
Handle case for empty ROOT file

An empty ROOT files contains no ProcessHistory which causes problems when filling the ProductRegistry at construction time. Now skip empty files until we find a file with data.

Dr15Jones committed:
Set no-process lookup at begin job time

Instead of dynamically determining what data product to get at Run/LuminosityBlock/Event transition time, the data product to get is now determined at begin job time. This is done for no-process and the skip current process cases.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45913

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/test/Event_t.cpp  modified in PR(s): #48689
  - File  FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/eventprincipal_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/generichandle_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Integration/plugins/PutOrMergeTestSource.cc  modified in PR(s): #48689
  - File  FWCore/Integration/test/testProcessBlockNOMergeOfMergedFiles_cfg.py  modified in PR(s): #48794
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #48689

Code check has found code style and quality issues which could be resolved by applying following patch(s)

- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45913/code-checks.patch
e.g. `curl -k https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45913/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

Dr15Jones committed:
apply result of code checks

Dr15Jones committed:
Define skippCurrentProcessLabel in one place

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45914

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/test/Event_t.cpp  modified in PR(s): #48689
  - File  FWCore/Framework/test/event_getrefbeforeput_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/eventprincipal_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/generichandle_t.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Framework/test/productregistry.cppunit.cc  modified in PR(s): #48689
  - File  FWCore/Integration/plugins/PutOrMergeTestSource.cc  modified in PR(s): #48689
  - File  FWCore/Integration/test/testProcessBlockNOMergeOfMergedFiles_cfg.py  modified in PR(s): #48794
  - File  FWCore/TestProcessor/src/TestProcessor.cc  modified in PR(s): #48689

Dr15Jones committed:
Add back in commented out class

Dr15Jones commented:
test parameters:
  - full_cmssw = true

Dr15Jones commented:
please test

makortel commented on DataFormats/StdDictionaries/src/classes_def_others.xml:
```diff
@@ -25,5 +25,5 @@
  <class name="std::atomic<int>"/>
  <class name="std::__atomic_base<int>"/>
- <class name="std::monostate"/>
+<class name="std::monostate"/>
```
Could this (whitespace-only) change be removed?

makortel commented on FWCore/Framework/src/DroppedDataProductResolver.h:
```diff
@@ -13,18 +13,16 @@ namespace edm {
   public:
     DroppedDataProductResolver(std::shared_ptr<ProductDescription const> bd)
-        : ProductResolverBase(), m_provenance(std::move(bd), {}) {}
+        : ProductResolverBase(), m_provenance(bd, {}) {}
```
I see this change was from `clang-tidy`.

Dr15Jones replied:
yes

makortel commented on FWCore/Framework/src/EDConsumerBase.cc:
```diff
@@ -268,7 +270,7 @@ ProductResolverIndexAndSkipBit EDConsumerBase::indexFromIfExactMatch(EDGetToken
     }
   }
-  return ProductResolverIndexAndSkipBit(edm::ProductResolverIndexInvalid, false);
+  return ProductResolverIndexAndSkipBit(edm::ProductResolverIndexInitializing, false);
```
The reason for this change is not immediately clear to me.

Dr15Jones replied:
It is now possible for info.m_index to return `ProductResolverIndexInvalid` (e.g. if told to `@skipCurrentProcess` and that was determined to match no data product). As such, I needed a different denotation that there is a problem where the token was somehow not properly initialized.

makortel replied:
Ok

makortel commented on FWCore/Framework/src/ProductResolvers.cc:
```diff
@@ -881,402 +845,7 @@ namespace edm {
-        << "SingleChoiceNoProcessProductResolver::connectTo() not implemented and should never be called.\n"
-        << "Contact a Framework developer\n";
-  }
-
```
Nice :)

makortel reviewed (commented):
As we discussed in person, we should probably wait for 16_0_X before merging this one

cmsbuild commented:
-1

**Failed Tests**:   ClangBuild 
**Size**: This PR adds an extra 476KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ddf0c2/47873/summary.html
**COMMIT**: 0ee576f04a5f3932d24a8d13345bb9248a2341b6
**CMSSW**: CMSSW_15_1_X_2025-08-26-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48807/47873/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @CMSTrackerDPG cms-sw/cmssw#48796

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ddf0c2/47873/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ddf0c2/47873/git-merge-result


## Clang Build

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' /usr/bin/time -v scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

Dr15Jones committed:
Removed unused variables

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45923

Dr15Jones committed:
Improve exception message from GenParticleProducer

Dr15Jones committed:
Conditionally produce/consume in GeneratorSmearedProducer

Only call produces and consumes if the data products the module
want appear to be in the job.
This solves the case where the module has been added to a job
just on the remote possibility it might be needed but infact
the values which are actually wanted come from the source.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45925

Dr15Jones committed:
Call ProductRegistry::setProcessOrder from sources

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45926

Dr15Jones committed:
fixed indention

Dr15Jones committed:
Improved code clarity based on code review

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/45927

cmsbuild commented:
Pull request #48807 was updated. @Dr15Jones, @bbilin, @lviliani, @makortel, @menglu21, @mkirsano, @smuzaffar, @tvami can you please check and sign again.

Dr15Jones committed:
Fix getting data from MixingModule

The MixingModule does not use consumes when getting data from the
mixing files.

cmsbuild commented:
Pull request #48807 was updated. @Dr15Jones, @bbilin, @civanch, @kpedro88, @lviliani, @makortel, @mdhildreth, @menglu21, @mkirsano, @smuzaffar, @tvami can you please check and sign again.

cmsbuild commented:
+1

**Size**: This PR adds an extra 36KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ddf0c2/47899/summary.html
**COMMIT**: 5552a3e90a0bd617eb1d7b3e5f34981d78a87cc0
**CMSSW**: CMSSW_15_1_X_2025-08-28-1100/el8_amd64_gcc12
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/48807/47899/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - You potentially [removed 171 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+ddf0c2/70041/validateJR/logRootQA-events.log) from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - [Reco comparison results: 0 differences found in the comparisons](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+ddf0c2/70041/validateJR.html)
 - DQMHistoTests: Total files compared: 50
 - DQMHistoTests: Total histograms compared: 4115479
 - DQMHistoTests: [Total failures: 25](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+ddf0c2/70041/validateJR.html)
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 4115434
 - DQMHistoTests: Total skipped: 20
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 49 files compared)
 - Checked 215 log files, 184 edm output root files, 50 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_15_1_X_2025-08-28-1100+ddf0c2/70041/triggerResults): no differences found

makortel commented:
+core

cmsbuild commented:
Pull request has been put on hold by  @makortel
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

cmsbuild commented:
Milestone for this pull request has been moved to CMSSW_16_0_X. Please open a backport if it should also go in to CMSSW_15_1_X.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-48807/46046

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/Principal.cc  modified in PR(s): #48816
  - File  FWCore/Framework/src/ProductResolvers.cc  modified in PR(s): #48816
  - File  FWCore/Framework/src/ProductResolvers.h  modified in PR(s): #48816
  - File  FWCore/Framework/src/ProductResolversFactory.cc  modified in PR(s): #48816

cmsbuild commented:
Pull request #48807 was updated. @civanch, @kpedro88, @lviliani, @mdhildreth, @mkirsano, @ssen, @theofil can you please check and sign again.

lviliani commented:
+generators
